target/ppc/translate/dfp-impl.c.inc

```diff
@@ -7,226 +7,223 @@ static inline TCGv_ptr gen_fprp_ptr(int reg)
     return r;
 }
 
-#define GEN_DFP_T_A_B_Rc(name)                   \
-static void gen_##name(DisasContext *ctx)        \
-{                                                \
-    TCGv_ptr rd, ra, rb;                         \
-    if (unlikely(!ctx->fpu_enabled)) {           \
-        gen_exception(ctx, POWERPC_EXCP_FPU);    \
-        return;                                  \
-    }                                            \
-    gen_update_nip(ctx, ctx->base.pc_next - 4);  \
-    rd = gen_fprp_ptr(rD(ctx->opcode));          \
-    ra = gen_fprp_ptr(rA(ctx->opcode));          \
-    rb = gen_fprp_ptr(rB(ctx->opcode));          \
-    gen_helper_##name(cpu_env, rd, ra, rb);      \
-    if (unlikely(Rc(ctx->opcode) != 0)) {        \
-        gen_set_cr1_from_fpscr(ctx);             \
-    }                                            \
-    tcg_temp_free_ptr(rd);                       \
-    tcg_temp_free_ptr(ra);                       \
-    tcg_temp_free_ptr(rb);                       \
+#define TRANS_DFP_T_A_B_Rc(NAME)                             \
```

Comment:
Not specific to this code, but a general thought: I wonder if we should turn macros like this in methods that receive a function pointer, e.g.:
```c
static bool do_dfp_t_a_b_rc(DisasContext *ctx, arg_X_rc *a,
                    void (*gen_helper)(TCGv_ptr, TCGv_ptr, TCGv_ptr, TCGv_ptr))
```
and then use the `TRANS` macro
```c
TRANS(DADD, do_dfp_t_a_b_rc, gen_helper_DADD)
TRANS(DADDQ, do_dfp_t_a_b_rc, gen_helper_DADDQ)
TRANS(DSUB, do_dfp_t_a_b_rc, gen_helper_DSUB)
TRANS(DSUBQ, do_dfp_t_a_b_rc, gen_helper_DSUBQ)
```
This way we'd show more info in the code (all these instruction use the same argument struct, the helper expects these type, etc.) and might be easier to debug with GDB.

Reply:
I was halfway through doing this, but then, after noting that, for some cases, the arg_xxx structure is different between instructions that use the same macro. This is the case for the instructions that use the ``TRANS_DFP_T_B_U32_U32_Rc`` and ``TRANS_DFP_T_FPR_I32_Rc`` macros.

So decided not worth doing for DFP.

Reply:
That's the kind of thing that makes these macros confusing. For these cases, I would break it into three methods. E.g.: `do_dfp_t_b_u32_u32_rc_Z23_tb` and `do_dfp_t_b_u32_u32_rc_Z23_te_tb` that call `do_dfp_t_b_u32_u32_rc`. The more methods we have until we call the helper, however, the less likely the compiler will inline everything.

But then again, this was a general thought, it might not be worth it in this case. That's a lot of instructions being moved to decodetree, and I suppose most macros with more than one parameter will have this kind of problem.